### PR TITLE
Ensure external refreshes will also refresh internal searcher to minimize segment creation

### PR DESCRIPTION
**Pull request by s1monw**
We cut over to internal and external IndexReader/IndexSeacher in #26972 which uses
two independent searcher managers. This has the downside that refreshes of the external
reader will never clear the internal version map which in-turn will trigger additional
and potentially unnecessary segment flushes since memory must be freed. Under heavy
indexing load with low refresh intervals this can cause excessive segment creation which
causes high GC activity and significantly increases the required segment merges.

This change adds a dedicated external reference manager that delegates refreshes to the
internal reference manager that then `steals` the refreshed reader from the internal
reference manager for external usage. This ensures that external and internal readers
are consistent on an external refresh. As a sideeffect this also releases old segments
referenced by the internal reference manager which can potentially hold on to already merged
away segments until it is refreshed due to a flush or indexing activity.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -224,8 +224,9 @@ public InternalEngine(EngineConfig engineConfig) {
                     throw e;
                 }
             }
-            internalSearcherManager = createSearcherManager(new SearcherFactory(), false);
-            externalSearcherManager = createSearcherManager(new SearchFactory(logger, isClosed, engineConfig), true);
+            externalSearcherManager = createSearcherManager(new SearchFactory(logger, isClosed,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: nit: this fits on one line. yay.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -248,6 +249,75 @@ public InternalEngine(EngineConfig engineConfig) {
...
+     * since no indexing is happening and refreshes are only happening to the external reader manager, while with
+     * this specialized implementation an external refresh will immediately be reflected on the internal reader
+     * and old segments can be released in the same way previous version did this (as a side-effect of _refresh)
+     */
+    @SuppressForbidden(reason = "reference counting is required here")
+    private static final class ExternalSearcherManager extends ReferenceManager<IndexSearcher> {
+        private final SearcherFactory searcherFactory;
+        private final SearcherManager internalSearcherManager;
+
+        ExternalSearcherManager(SearcherManager manager, SearcherFactory searcherFactory) throws IOException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: nit: manger -> internalManager

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -456,18 +526,17 @@ private String loadOrGenerateHistoryUUID(final IndexWriter writer, boolean force
         return uuid;
     }
 
-    private SearcherManager createSearcherManager(SearcherFactory searcherFactory, boolean readSegmentsInfo) throws EngineException {
+    private ExternalSearcherManager createSearcherManager(SearchFactory factory) throws EngineException {
         boolean success = false;
         SearcherManager searcherManager = null;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: nit: searcherManager -> internalSearcherManager

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -456,18 +526,17 @@ private String loadOrGenerateHistoryUUID(final IndexWriter writer, boolean force
         return uuid;
     }
 
-    private SearcherManager createSearcherManager(SearcherFactory searcherFactory, boolean readSegmentsInfo) throws EngineException {
+    private ExternalSearcherManager createSearcherManager(SearchFactory factory) throws EngineException {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: nit: factory -> externalSearcherFactory

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: also, good that this boolean is goon.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 5]
```diff
@@ -4226,7 +4226,7 @@ public void assertSameReader(Searcher left, Searcher right) {
         List<LeafReaderContext> rightLeaves = ElasticsearchDirectoryReader.unwrap(right.getDirectoryReader()).leaves();
         assertEquals(rightLeaves.size(), leftLeaves.size());
         for (int i = 0; i < leftLeaves.size(); i++) {
-            assertSame(leftLeaves.get(i).reader(), rightLeaves.get(0).reader());
+            assertSame(leftLeaves.get(i).reader(), rightLeaves.get(i).reader());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: sneaky :)

**Review by bleskes — APPROVED** [review]
LGTM. Left some nits that would have helped me while reading this PR. feel free to do with them as you please.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 5 above)
Comment (s1monw) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: yeah suddenly we have a test that triggered it 💃

**Comment by s1monw** [comment]
@elasticmachine test this please

**Comment by jasontedor** [comment]
@s1monw It appears the "backport pending" label can be removed now?

**Comment by bleskes** [comment]
@jasontedor agreed. I removed the backport pending label.

**Pull request merged**